python/pyarrow/interchange/from_dataframe.py

```diff
@@ -0,0 +1,590 @@
+                               allow_copy)
+
+    if null_kind == ColumnNullType.USE_SENTINEL:
+        if not allow_copy:
+            raise RuntimeError(
+                "PyArrow Array will always be created out of a NumPy ndarray "
+                "and a copy is required which is forbidden by allow_copy=False"
+            )
+        bytemask = [value == sentinel_val for value in indices.to_pylist()]
+        indices = pa.array(indices.to_pylist(), mask=bytemask)
+
+    dict_array = pa.DictionaryArray.from_arrays(indices, dictionary)
+    return dict_array, buffers
+
+
+def datetime_column_to_array(
+    col: ColumnObject,
+    allow_copy: bool = True
+) -> tuple[pa.Array, Any]:
+    """
+    Convert a column holding DateTime data to a NumPy array.
+
+    Parameters
+    ----------
+    col : ColumnObject
+
+    Returns
+    -------
+    tuple
+        Tuple of pa.Array holding the data and the memory owner object
+        that keeps the memory alive.
+    """
+    buffers = col.get_buffers()
+    data_buff, data_type = buffers["data"]
+    try:
+        validity_buff, validity_dtype = buffers["validity"]
+    except TypeError:
+        validity_buff = None
+
+    format_str = data_type[2]
+    unit, tz = parse_datetime_format_str(format_str)
+    data_dtype = pa.timestamp(unit, tz=tz)
+
+    # Data buffer
+    data_pa_buffer = pa.foreign_buffer(data_buff.ptr, data_buff.bufsize)
+
+    # Validity buffer
+    validity_pa_buff = validity_buff
+    bytemask = None
+    if validity_pa_buff:
+        validity_pa_buff, bytemask = validity_buffer(validity_buff,
+                                                     validity_dtype,
+                                                     col.describe_null,
+                                                     col.offset)
+
+    # Constructing a pa.Array from data and validity buffer
+    array = pa.Array.from_buffers(
+        data_dtype,
+        col.size(),
+        [validity_pa_buff, data_pa_buffer],
+        offset=col.offset,
+    )
+
+    null_kind, sentinel_val = col.describe_null
+    if null_kind == ColumnNullType.USE_SENTINEL:
+        # pandas iNaT
+        if sentinel_val == -9223372036854775808:
+            bytemask = np.isnan(np.array([str(e) for e in array.to_pylist()],
+                                dtype="datetime64[ns]"))
+        else:
+            raise NotImplementedError(
+                f"Sentinel value {sentinel_val} for datetime is not yet "
+                "supported.")
+
+    # In case a bytemask was constructed with validity_buffer() call
+    # or with sentinel_value then we have to add the mask to the array
+    if bytemask is not None:
+        if not allow_copy:
+            raise RuntimeError(
+                "PyArrow Array will always be created out of a NumPy ndarray "
+                "and a copy is required which is forbidden by allow_copy=False"
+            )
+        return pa.array(array.to_pylist(), mask=bytemask), buffers
```

Comment:
We should make this easier to create a new array from an existing array + new validity bitmap or mask, without the need to go through `to_pylist`. https://issues.apache.org/jira/browse/ARROW-7071 might be a related feature request.

Although in this case, doing `array.to_numpy()` might be OK as well? (conversion to numpy should be cheaper than to a python list)